src/lib/currency/currency.ml

```diff
@@ -66,6 +66,32 @@ end = struct
   [%%define_locally
   Unsigned.(to_uint64, of_uint64, of_int, to_int, of_string, to_string)]
 
+  let precision = 9
+
+  let precision_exp = Unsigned.of_int 1_000_000_000
+
+  let to_formatted_string amount =
+    let whole = Unsigned.div amount precision_exp in
+    let remainder = Unsigned.rem amount precision_exp in
+    to_string whole ^ "." ^ to_string remainder
+
+  let of_formatted_string input =
+    let parts = String.split ~on:'.' input in
+    match parts with
+    | [whole] ->
+        of_string (whole ^ String.make precision '0')
+    | [whole; decimal] ->
+        let decimal_length = String.length decimal in
+        if
+          let open Int in
+          decimal_length > precision
+        then of_string (whole ^ String.sub decimal ~pos:0 ~len:precision)
+        else
+          of_string
+            (whole ^ decimal ^ String.make (precision - decimal_length) '0')
+    | [] | _ ->
```

Comment:
would be nice to put in the name of the function in the error message, "of_formatted_sring: ..."